[Dynamo]Expose bytecode hooks and add example usage for decompilation in docs

youkaichao opened the pull request:
Dynamo dynamically translate bytecode of python functions, which is powerful but with difficult-to-understand bytecode. Most users cannot understand python bytecode. Although a general purpose way to decompile python bytecode into source code is very difficult, I find that this work can be greatly simplified since Dynamo already cleans up the code: the bytecode generated by Dynamo is a reduced subset of well-behaved python bytecode.

I created a tiny decompiler for pytorch 2.0, named `depyf`: https://github.com/youkaichao/depyf .

There are several takeaways:

- **It supports pyton 3.7 - 3.11 (both inclusive), the same python versions supported by pytorch.** Since the main usage of this library is to understand pytorch 2.0, I plan to keep pace with pytorch. If pytorch supports a new python version, I can add support for that. (Actually, the core code is just about 1k lines. Adding support for new versions of python bytecode can be done in just several days.)
- **I have tested the correctness of decompiled source code in torchbench.** I capture the modified bytecode generated by Dynamo, decompile it into source code, and then compile it into new bytecode, replace the Dynamo generated bytecode with new bytecode. And **it passed all the accuracy tests for timm models**. For huggingface models, the situation is more complicated: all failed cases are caused by the compile step: some functions use the `__class__`  as closure variables, but decompiler can only get the code object, so it has no way to figure out the `__class__` , leading to a name error when compiling the decompiled code. That said, it passed the rest tests without the `__class__` issue. Please see the log file https://cloud.tsinghua.edu.cn/f/685e4af8d930499baa7c/?dl=1 and https://cloud.tsinghua.edu.cn/f/cab89500e15e4b62890b/?dl=1 for details.

With the above efforts, I think it would be great to add an additional logging option in Dynamo: we can try to decompile the generated bytecode into source code, so that users can have a rough idea of what the modified bytecode does. It does not affect the workflow of Dynamo, but just adds more debug information.

An example code from the [doc](https://pytorch.org/docs/main/torch.compiler_deepdive.html):

```python
from typing import List
import torch
from torch import _dynamo as torchdynamo
def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
    print("my_compiler() called with FX graph:")
    gm.graph.print_tabular()
    return gm.forward  # return a python callable

@torchdynamo.optimize(my_compiler)
def toy_example(a, b):
    x = a / (torch.abs(a) + 1)
    if b.sum() < 0:
        b = b * -1
    return x * b
for _ in range(100):
    toy_example(torch.randn(10), torch.randn(10))
```

Run with `export TORCH_LOGS="+dynamo,guards,bytecode"`.

Bytecode logging:

```
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] ORIGINAL BYTECODE toy_example /Users/youkaichao/DeepLearning/depyf/ykc_test.py line 8 
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  10           0 LOAD_FAST                0 (a)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               2 LOAD_GLOBAL              0 (torch)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               4 LOAD_METHOD              1 (abs)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               6 LOAD_FAST                0 (a)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               8 CALL_METHOD              1
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              10 LOAD_CONST               1 (1)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              12 BINARY_ADD
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              14 BINARY_TRUE_DIVIDE
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              16 STORE_FAST               2 (x)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  11          18 LOAD_FAST                1 (b)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              20 LOAD_METHOD              2 (sum)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              22 CALL_METHOD              0
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              24 STORE_FAST               3 (__temp_2)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  12          26 LOAD_FAST                3 (__temp_2)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              28 LOAD_CONST               2 (0)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              30 COMPARE_OP               0 (<)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              32 POP_JUMP_IF_FALSE       21 (to 42)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  13          34 LOAD_FAST                1 (b)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              36 LOAD_CONST               3 (-1)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              38 BINARY_MULTIPLY
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              40 STORE_FAST               1 (b)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  14     >>   42 LOAD_FAST                2 (x)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              44 LOAD_FAST                1 (b)
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              46 BINARY_MULTIPLY
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              48 RETURN_VALUE
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,929] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] MODIFIED BYTECODE toy_example /Users/youkaichao/DeepLearning/depyf/ykc_test.py line 8 
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]   8           0 LOAD_GLOBAL              3 (__compiled_fn_0)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               2 LOAD_FAST                0 (a)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               4 LOAD_FAST                1 (b)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               6 CALL_FUNCTION            2
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               8 UNPACK_SEQUENCE          2
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              10 STORE_FAST               2 (x)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              12 POP_JUMP_IF_FALSE       12 (to 24)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              14 LOAD_GLOBAL              4 (__resume_at_34_1)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              16 LOAD_FAST                1 (b)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              18 LOAD_FAST                2 (x)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              20 CALL_FUNCTION            2
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              22 RETURN_VALUE
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]         >>   24 LOAD_GLOBAL              5 (__resume_at_42_2)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              26 LOAD_FAST                1 (b)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              28 LOAD_FAST                2 (x)
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              30 CALL_FUNCTION            2
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              32 RETURN_VALUE
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 23:56:44,930] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]
```

New output with this PR:

```
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] possible source code:
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] def toy_example(a, b):
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]     __temp_1 = __compiled_fn_0(a, b)
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]     x = __temp_1[0]
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]     if __temp_1[1]:
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]         return __resume_at_34_1(b, x)
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]     return __resume_at_42_2(b, x)
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,535] [0/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] If you find the decompiled code is wrong,please submit an issue at https://github.com/youkaichao/depyf/issues.
```

The rest two log (please pay attention to the output `possible source code:`):

```
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] ORIGINAL BYTECODE <resume in toy_example> /workspace/youkaichao/code/pytorch/ykc.py line 12 
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  12           0 JUMP_ABSOLUTE           22 (to 44)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               2 LOAD_FAST                2 (a)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               4 LOAD_GLOBAL              0 (torch)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               6 LOAD_ATTR                1 (abs)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               8 LOAD_FAST                2 (a)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              10 CALL_FUNCTION            1
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              12 LOAD_CONST               1 (1)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              14 BINARY_ADD
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              16 BINARY_TRUE_DIVIDE
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              18 STORE_FAST               1 (x)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              20 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              22 LOAD_ATTR                2 (sum)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              24 CALL_FUNCTION            0
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              26 STORE_FAST               3 (__temp_2)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              28 LOAD_FAST                3 (__temp_2)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              30 LOAD_CONST               2 (0)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              32 COMPARE_OP               0 (<)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              34 POP_JUMP_IF_FALSE       22 (to 44)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              36 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              38 LOAD_CONST               3 (-1)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              40 BINARY_MULTIPLY
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              42 STORE_FAST               0 (b)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  14     >>   44 LOAD_FAST                1 (x)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              46 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              48 BINARY_MULTIPLY
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              50 RETURN_VALUE
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] MODIFIED BYTECODE <resume in toy_example> /workspace/youkaichao/code/pytorch/ykc.py line 12 
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  12           0 LOAD_GLOBAL              3 (__compiled_fn_3)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               2 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               4 LOAD_FAST                1 (x)
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               6 CALL_FUNCTION            2
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               8 UNPACK_SEQUENCE          1
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              10 RETURN_VALUE
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,566] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,567] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] possible source code:
[2023-10-06 16:25:21,567] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] def <resume in toy_example>(b, x):
[2023-10-06 16:25:21,567] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]     return __compiled_fn_3(b, x)[0]
[2023-10-06 16:25:21,567] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,567] [1/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] If you find the decompiled code is wrong,please submit an issue at https://github.com/youkaichao/depyf/issues.
```

```
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] ORIGINAL BYTECODE <resume in toy_example> /workspace/youkaichao/code/pytorch/ykc.py line 12 
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  12           0 JUMP_ABSOLUTE           18 (to 36)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               2 LOAD_FAST                2 (a)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               4 LOAD_GLOBAL              0 (torch)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               6 LOAD_ATTR                1 (abs)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               8 LOAD_FAST                2 (a)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              10 CALL_FUNCTION            1
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              12 LOAD_CONST               1 (1)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              14 BINARY_ADD
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              16 BINARY_TRUE_DIVIDE
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              18 STORE_FAST               1 (x)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              20 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              22 LOAD_ATTR                2 (sum)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              24 CALL_FUNCTION            0
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              26 STORE_FAST               3 (__temp_2)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              28 LOAD_FAST                3 (__temp_2)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              30 LOAD_CONST               2 (0)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              32 COMPARE_OP               0 (<)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              34 POP_JUMP_IF_FALSE       22 (to 44)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  13     >>   36 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              38 LOAD_CONST               3 (-1)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              40 BINARY_MULTIPLY
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              42 STORE_FAST               0 (b)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  14     >>   44 LOAD_FAST                1 (x)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              46 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              48 BINARY_MULTIPLY
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              50 RETURN_VALUE
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,579] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] MODIFIED BYTECODE <resume in toy_example> /workspace/youkaichao/code/pytorch/ykc.py line 12 
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]  12           0 LOAD_GLOBAL              3 (__compiled_fn_4)
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               2 LOAD_FAST                0 (b)
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               4 LOAD_FAST                1 (x)
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               6 CALL_FUNCTION            2
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]               8 UNPACK_SEQUENCE          1
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]              10 RETURN_VALUE
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] possible source code:
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] def <resume in toy_example>(b, x):
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG]     return __compiled_fn_4(b, x)[0]
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] 
[2023-10-06 16:25:21,580] [2/0] torch._dynamo.convert_frame.__bytecode: [DEBUG] If you find the decompiled code is wrong,please submit an issue at https://github.com/youkaichao/depyf/issues.
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

youkaichao committed:
use depyf for logging source code of modified bytecode

youkaichao commented:
cc @jansel

williamwen42 reviewed (commented):
This could have some utility for me when debugging dynamo-generated bytecode. However, I think it's only helpful to see decompiled bytecode/decompilation error when the generated bytecode is wrong (e.g. causes a segfault because I pop too many things), but I'm not sure how your library will display that, especially since there probably isn't any python program that could generate bytecode that segfaults. Another point I'm concerned about is if `depyf` can decompile continuation functions - these functions can't be compiled from valid Python source since they contain things like absolute jumps, i.e. `goto`-equivalent behavior. Continuation functions with contexts I think are particularly tricky.

At the very least, place this under a new logging artifact and don't make `depyf` a required dependency, i.e., don't error when `depyf` is not present.

jansel reviewed (changes requested):
I agree with the comments above.  It also looks like tests are failing due to the hard requirement.

Perhaps this could be done more cleanly with something like
```
torch._dynamo.config.output_bytecode_hook = ...
```

Then here we just do:
```
if config.output_bytecode_hook:
   config.output_bytecode_hook(code, out_code)
```

This would allow most of this code to go in the depyf repo and could also possibly be used for other purposes.

youkaichao commented:
> Another point I'm concerned about is if `depyf` can decompile continuation functions

@williamwen42 `depyf` is specially designed to work with continuation functions. You can see the output above. It works by dead bytecode elimination.

> it's only helpful to see decompiled bytecode/decompilation error when the generated bytecode is wrong

@williamwen42 the decompilation will raise error like `pop from empty list`. And the traceback will let you know which instruction caused the error. By debugging into the `depyf`, you can get much more information.

> place this under a new logging artifact and don't make depyf a required dependency

Yeah, that can be done.

> output_bytecode_hook

@jansel the `output_bytecode_hook` seems good. I will have a try.

youkaichao committed:
place depyf under bytecode_src logger

youkaichao committed:
use output_bytecode_hooks

youkaichao committed:
add bytecode_src_hooks

youkaichao commented:
The functionality is now controlled by an additional logging site `export TORCH_LOGS="+bytecode_src"`.

I also enable the `output_bytecode_hooks` to modify the output bytecode (and this can be used to easily replace bytecode with decompile--> compile new bytecode) if it returns a new codeobject. Anyway, I think it makes no harm to introduce more functionality. I'm open to restrict this functionality (e.g. the hook must return `None`), if you want.

malfet reviewed (changes requested):
I don't think this is a right idea to prefer one decompiler to another. As @jansel suggested we should perhaps have a hooks that any projects can register into and then they'll be called.

I know that https://github.com/rocky/python-uncompyle6 has not been updated to support Python-3.11, but otherwise it looks like a much more mature decompiler than `depyf`, so it feels a bit strange that one is picked over another, but I would be more inclined to accept it if PR description could be expanded to cover the difference in code output/perf across various decompilers

youkaichao commented:
> I know that [rocky/python-uncompyle6](https://github.com/rocky/python-uncompyle6) has not been updated to support Python-3.11, but otherwise it looks like a much more mature decompiler than `depyf`.

`python-uncompyle6` and related project like `decompyle3` are what I have tried the first place. However, they just don't work for Dynamo ...... They aim for decompiling `.pyc` files at the whole source code file level. Their output is a `.py` file that can be executed individually. Dynamo generated bytecode cannot be conviently converted into `.pyc` files because that requires much more than the bytecode. Say the bytecode referenced a free variable `x` from the outer closure, `python-uncompyle6`/`decompyle3` need to dump the content of `x` into source code, while `depyf` just needs to use the variable name `x`.

Another point is that the code objects produced by Dynamo are not always "valid" in python. Resume functions have function name `<resume in xxx>`, which is invalid python function name, and is unexpected for normal decompilers. `depyf` takes care of it.

**Decompiling Dynamo bytecode has so many unique requirements that cannot be achieved by general purpose decompilers. That's why I created depyf.**

Hope it helps. @malfet

youkaichao committed:
do not error if depyf is not installed

youkaichao committed:
remove default bytecode hooks

youkaichao commented:
Thanks for the feedback. I removed `depyf` as the default decompiler. This PR just add hooks for bytecode. I also updated `depyf` to have `depyf.install()` to enable the debug logging. Later I will update the doc at https://pytorch.org/docs/main/torch.compiler_deepdive.html to show users an example of how to use this feature.

jansel commented on torch/_dynamo/config.py:
```diff
@@ -352,6 +352,8 @@ def is_fbcode():
 _experimental_support_context_fn_in_torch_utils_checkpoint = False
 
 
+output_bytecode_hooks = []
```
Add a docstring above this explaining what it does

youkaichao replied:
Sure, will do.

jansel commented on torch/_dynamo/types.py:
```diff
@@ -94,3 +94,10 @@ def __call__(
         ...
+
+
+class BytecodeHook(Protocol):
```
unused?

youkaichao replied:
It is used to specify the protocol. Since no hooks are installed by default, we have to show users the protocol of hooks.

youkaichao committed:
update doc

youkaichao committed:
move hooks to convert_frame

youkaichao committed:
update doc

youkaichao committed:
fix doc format

youkaichao committed:
fix doc

williamwen42 commented:
How do you plan on supporting new Python versions? I'm planning on looking at 3.12 support soon. A big use case to me for using a decompiler would be for new Python versions, since otherwise, we don't touch bytecode generation very much.

ezyang commented:
I have a minor nit about not allowing the hook to change the output bytecode, but otherwise this looks good, the modification is nice.

@williamwen42 what exactly do you mean?

williamwen42 commented:
@ezyang I think a nice use case for a decompiler is to use it as a debugging tool when supporting dynamo for future python versions. But my understanding for `depyf` is that there's a lot of hardcoded behavior dependent on the python version and on dynamo behavior.

youkaichao commented on torch/_dynamo/convert_frame.py:
```diff
@@ -534,6 +541,11 @@ def log_bytecode(prefix, name, filename, line_no, code):
+        for hook in output_bytecode_hooks:
+            hook_output = hook(code, out_code)
+            if hook_output is not None:
+                out_code = hook_output
```
I use this functionality to test the correctness of decompiler. My workflow: bytecode ==decompiler==> source code ==compiler==> new bytecode, replace bytecode with new bytecode and run dynamo benchmark tests against accuracy.

ezyang replied:
Does this always work? Why don't we write our Dynamo bytecode stuff as plain Python 🤔

youkaichao replied:
What do you mean by "plain Python"? Can you elaborate on this sentence?

My point is that adding additional functionality wouldn't hurt anyway, and the hook developers risk themselves when modifying the bytecode.

ezyang replied:
What I mean is, I wonder if we should even be writing our reconstruction code as bytecode manually, maybe we could have just written Python and compiled it into bytecode...

youkaichao replied:
That's actually what I did. I think there is some misunderstanding. Talking to you in slack then.

youkaichao replied:
I find that you muted slack notification, so I place the comment here.

Below is my hook for testing the correctness of decompiler:

```python
import depyf
def pytorch_bytecode_replace_hook(code: types.CodeType, new_code: types.CodeType):

    try:
        decompiled_src = depyf.Decompiler(new_code).decompile(overwite_fn_name="f")
        scope = {}
        exec(decompiled_src, scope)
        return scope["f"].__code__
    except Exception as e:
        print("Decompilation fails due to: %s" % str(e))
```

The code attribute in Dynamo CacheEntry is read-only (not have a setter):

https://github.com/pytorch/pytorch/blob/80ea8784f392959f8d61e4cd37420819d5694bf1/torch/csrc/dynamo/eval_frame.c#L337C7

So the only way to replace the bytecode with new bytecode is in the hook side, before we return the cache entry to the C function.

In addition, I overwrite the generated function name to avoid invalid function name issue of resume functions.

youkaichao commented:
@williamwen42 I added a preliminary support for Python 3.12 just now. Not very complicated when my goal is to decompile into human readable source code. Please ping me if you have any questions.

Note that I implement bytecode described in https://docs.python.org/3.12/library/dis.html . Some opnames are not documented there, so I don't know how to implement them, e.g. INSTRUMENTED_XXX .

youkaichao commented:
I think there are two usecases for `depyf`:
- Help pytorch developers to support dynamo for future python versions.
- Help pytorch users to understand the generated bytecode.

youkaichao commented:
Hi, team, any further comment?

jansel commented on torch/_dynamo/convert_frame.py:
```diff
@@ -412,6 +413,12 @@ def maybe_cprofile(func):
     return func
 
 
+# Hooks for bytecode generated by Dynamo. The hook can do some
+# logging, as well as return a new code object to be used. Please
+# refer to `BytecodeHook` for the hook signature.
+output_bytecode_hooks: List[BytecodeHook] = []
```
Put this in config.py so it can be changed via `torch.compile(options=...)`

youkaichao replied:
Can you elaborate on how this can work out? I don't know `options` in `torch.compile` can go so deeply and affect these hooks. My understanding is that `options` are some pre-defined options, and the type signature is `Optional[Dict[str, Union[str, int, bool]]]`. Since we don't provide any set of predefined hooks, I don't know how to support hooks in `torch.compile(options=...)`.

jansel replied:
Ah right.  I think options only effect _inductor.config, my mistake.  It is fine to leave it here.

youkaichao commented:
Do we need to add some api like `register_bytecode_hook` in `torch/_dynamo/convert_frame.py`? The ideal case would be to mimic the forward hook design in the nn module:

```python
# register hook
handle = torch._dynamo.convert_frame.register_bytecode_hook(hook)
# remove hook
hook.remove()
torch._dynamo.convert_frame.clear_bytecode_hook(hook)
```

That would need more discussion and design though, and might exceed the scope of this PR.

youkaichao commented:
`torch.utils.hooks.RemovableHandle` seems to be a good option. Please tell me if this API design is good to go.

youkaichao committed:
expose a hook registration api

youkaichao commented:
@jansel  I added an API to register hooks in [dea40bb](https://github.com/pytorch/pytorch/pull/110714/commits/dea40bbf489466c15eaa374c10fd10a1ea9457e1). Please check if it is good to go. I'm pretty open anyway. If you don't like it, I can revert the commit. It's just my thought that registing hooks should be an API of pytorch rather than out-of-tree library directly manipulating an internal variable in pytorch.

youkaichao committed:
Merge remote-tracking branch 'origin' into depyf

youkaichao commented:
@pytorchmergebot merge

youkaichao commented:
@pytorchbot label "release notes: dynamo"

youkaichao commented:
@pytorchmergebot merge

youkaichao commented:
This PR has been merged now. Please feel free to comment If you have any other concern, we can revert this PR at any time :)

jon-chuang commented on docs/source/torch.compiler_deepdive.rst:
```diff
@@ -109,6 +112,17 @@ If you want to understand better what TorchDynamo is doing, you can run your cod
+.. code-block:: python
+
+   import depyf
+   depyf.install()
```
Just a nit - calling install everytime is a bit counter intuitive. Might I suggest `depyf.init()`?

jon-chuang replied:
Is there also an overhead to installing? In which case having `depyf.disable()` would be good too.

youkaichao replied:
We have `depyf.uninstall()` function, too. You can refer to the repo https://github.com/thuml/depyf .

youkaichao replied:
It indeed adds some overheads to compilation. So it's mainly used for understanding and debugging.

jon-chuang replied:
Cool. Would still suggest `depyf.enable()`, `depyf.disable()` as the API instead as `install()/uninstall()` is confusing and might be confused with `pip install`-ing something.

youkaichao replied:
I would vote for `register_hook()/remove_hook()`. But changing this API requires another PR to change the doc. I used to request approval from @jansel , but it seems he is not available recently.

jon-chuang replied:
@youkaichao, hmm I see. I think the pytorch team is busy at pytorch conference https://pytorch.org/events. They should be more responsive by next week.

I think you can keep but deprecate the old API, and release the new API in a new release, and recommend new API changes in a PR in pytorch repo.

jansel replied:
This hasn't been released (in a stable branch) yet, so we are free to make changes.

youkaichao replied:
I'm thinking of refactoring the deep dive doc to something like https://github.com/thuml/depyf#example-usage . The dumped src is easy to understand and contains almost all the information of Dynamo. The dump src functionality is much more useful than just bytecode hooks. WDYT? @jansel @jon-chuang
